Adds a 'Getting Started with Eventing' guide to the docs (#1592)
knative/docs: docs/eventing/getting-started.md

---
title: "Getting Started with Eventing"
#linkTitle: "OPTIONAL_ALTERNATE_NAV_TITLE"
weight: 9
type: "docs"
---

This getting started guide will walk you through all the steps involved in a typical Knative workflow. You will learn how to create, send, and verify Knative events.

## Before you begin

You need:

- A Kubernetes cluster with [Knative Eventing installed](https://knative.dev/docs/install/index.html). 

> ## Set up components

Author: I've changed it to subcomponents globally.

Before you start to send Knative events, you need to create the components needed to move the events. In this guide, you will be creating components individually, but you can also create components by deploying a [single YAML file](https://raw.githubusercontent.com/akashrv/docs/qs/docs/eventing/samples/hello-world/quick-start.yaml).

> ### Configure namespace for eventing

Reviewer: Whats the bigger task/goal here? We say configuring a namespace, but what is the large "task" that the user should accomplish? I honestly dont know the answer here. But below, for example, it is mentioned that "creating a namespace also creates a 'broker'". We should follow up and find out if thats the primary reason, otherwise whats missing here is the "WHY". 

Ie. "Setting the default Eventing namespace to ...[explain more here to help user understand this concept]"

Use this as an example for all other individual "steps" and take advantage of that opportunity to "teach" about the "why create/config/blah...." a namespace/etc. So for example when or if you later need to "build on the concept of a Knative ....." (in this example the Knative Eventing namespace), then its easier and that base has been established.

Think of each step as a build block and take full advantage of the opportunity to explain in-depth, that related concept.

This guide uses the default `namespace`, so there is no need to create a new one.

1. Set up the `namespace` for Knative Eventing. To set up the `namespace`, add a label to your `namespace` with this command:

```sh
kubectl label namespace default knative-eventing-injection=enabled
```

This label triggers Knative to add `Service Accounts`, `Role Bindings`, and a `Broker` to your `namespace`. You'll learn more about `Brokers` in the next section.

> ### Examine broker for issues

Reviewer: Again, we are moving away from what the "users goal/task" is here. The user might need to actual perform the step of "examining the broker" but the real user task is "validating that the broker is running" or "ensuring that the broker is created". 

Its a slight difference but on top of that, mentioning "issues" this early (ie. getting started guide is focused at first time users) might cause concern

---
title: "Getting Started with Eventing"
#linkTitle: "OPTIONAL_ALTERNATE_NAV_TITLE"
weight: 9
type: "docs"
---

> This getting started guide will walk you through all the steps involved in a typical Knative workflow. You will learn how to create, send, and verify events in Knative.

Reviewer 2: Great job on adding more explanation! I think you addressed a lot of Richie's concerns.

> This getting started guide will walk you through all the steps involved in a typical Knative workflow. You will learn how to create, send, and verify events in Knative.

Reviewer 3: "workflow" is potentially an overloaded term in eventing. Could we say "in a typical Knative use-case"?

## Before you begin

You need:

- A Kubernetes cluster with [Knative Eventing installed](./install/index.html). 

> ## Setting up Eventing Subcomponents

Reviewer 2: "Subcomponents" shouldn't be capitalized

> Before you start to send events, you need to create the components needed to move the events. In this guide, you will be creating components individually, but you can also create components by deploying a [single YAML file](https://raw.githubusercontent.com/akashrv/docs/qs/docs/eventing/samples/hello-world/quick-start.yaml).

Reviewer 2: I'm not sure we should give folks the option of creating everything all at once. What if we re-wrote the last sentence to say, "This guide walks you through creating each Eventing subcomponent."

In the next section, you'll learn more about one of the subcompoents added to your namespace, the `Broker`.

### Validating that the `Broker` is running

The `Broker` ensures that every event sent by event producers is sent to the correct event consumers. While you created the `Broker` when you labeled your namespace as ready for eventing, it is important to verify that your `Broker` is working correctly. In this guide, you will use the default broker.

1. Use the following command to verify that the `Broker` is in a healthy state:

```sh
kubectl -n default get Broker default
```

This should return the following:

```sh
NAME      READY   REASON   HOSTNAME                                                           AGE
default   True             default-Broker.default.svc.cluster.local   1m
```

When the `Broker` is **READY**, it can begin to manage the events it receives.

2. If the **READY** column reads **False**, wait 2 minutes and repeat Step 1. If the **READY** column still reads **False**, see the [Debugging Guide](./debugging/README.md) to troubleshoot the issue.

Now your `Broker` is ready to manage your events.

### Creating event consumers

> These subcomponents receive the events sent by event producers (you'll create those a little later). You'll create two event consumers, `foo-display` and `bar-display`, so that you can see how to selectively send events to different consumers later.

Reviewer 3: Can we use more meaningful names here, I don't know what 'foo-display' and 'bar-display' are supposed to be doing. Maybe we have use a simple real-world example.

1. To create the `foo-display` subcomponent, enter the following command:

```sh
kubectl -n default apply -f - << END
apiVersion: apps/v1
kind: Deployment
metadata:
  name: foo-display
spec:
  replicas: 1
  selector:
    matchLabels: &labels
      app: foo-display
  template:
    metadata:
      labels: *labels
    spec:
      containers:
        - name: event-display

>           # Source code: https://github.com/knative/eventing-sources/blob/release-0.6/cmd/event_display/main.go

Reviewer 3: Link should be: https://github.com/knative/eventing-contrib/blob/release-0.6/cmd/event_display/main.go

>           image: gcr.io/knative-releases/github.com/knative/eventing-sources/cmd/event_display@sha256:37ace92b63fc516ad4c8331b6b3b2d84e4ab2d8ba898e387c0b6f68f0e3081c4

Reviewer 3: Does the image name also follow the renamed (event-contrib) as above?

Author: @akashrv Should we rename this?

Author: The image name (as of 7/26) still follows the old naming convention. I will check in on this as it is being published.

---

# Service pointing at the previous Deployment. This will be the target for event
# consumption.
kind: Service
apiVersion: v1
metadata:
  name: foo-display
spec:
  selector:
    app: foo-display
  ports:
  - protocol: TCP
    port: 80
    targetPort: 8080

> END

Reviewer 3: Why not use a ksvc here instead of a standard deployment?

Reviewer 2: @akashrv , @grantr , @adamharwayne  -- thoughts on this?

Reviewer 4: We totally could use a ksvc instead, but then that would also need to be a prerequisite. I have tended to using K8s svc + deployment to ease debugging, but I don't think this guide is worried about debugging, so it may not be relevant here. 

Please use @Harwayne instead of @adamharwayne.

Reviewer 3: I would expect that since Knative is mainly focused on enabling serving + eventing together, our guide would leverage Knative components when reasonable, especially for a getting started with eventing topic. Perhaps there's another topic we could write about using Knative Eventing with K8S deployments too?

Reviewer 2: Since this guide is focused on eventing, we've decided to keep it simple here and stick with a K8s service (after consulting with @akashrv ).

```

2. To create the `bar-display` subcomponent, enter the following command:

```sh
kubectl -n default apply -f - << END
apiVersion: apps/v1
kind: Deployment
metadata:
  name: bar-display
spec:
  replicas: 1
  selector:
    matchLabels: &labels
      app: bar-display
  template:
    metadata:
      labels: *labels
    spec:
      containers:
        - name: event-display
          # Source code: https://github.com/knative/eventing-sources/blob/release-0.6/cmd/event_display/main.go

>           image: gcr.io/knative-releases/github.com/knative/eventing-sources/cmd/event_display@sha256:37ace92b63fc516ad4c8331b6b3b2d84e4ab2d8ba898e387c0b6f68f0e3081c4

Reviewer 3: Same feedback as above with 'foo-service'

---

# Service pointing at the previous Deployment. This will be the target for event
# consumption.
kind: Service
apiVersion: v1
metadata:
  name: bar-display
spec:
  selector:
    app: bar-display
  ports:
  - protocol: TCP
    port: 80
    targetPort: 8080
END
```

3. Just like the `Broker`, verify that the event consumers are working with the following command:

```sh
kubectl -n default get deployments foo-display bar-display
```

This commmand shows all of the deployments that have been created in `foo-display` and `bar-display`. You should have one replica in each event consumer:

```sh
NAME           DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
foo-display    1         1         1            1           26s
NAME           DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
bar-display    1         1         1            1           16s
```

The number of replicas in the **DESIRED** column should match the number of replicas in the **AVAILABLE** column. This may take a few minutes. If after two minutes the numbers still do not match the example, then see the [Debugging Guide](./debugging/README.md).

### Creating `Triggers`

> `Triggers` allow events to register interest with a `Broker`. A `Trigger` is split into two parts: the `Filter`, which tracks interested events, and the `Subscriber`, which determines where the event should be sent. Triggers can search for various types of events:

Reviewer 3: "allow events to register interest..."  seems incorrect. Doesn't the trigger express a workload's interest in events it wants to receive?

1. To create the first `Trigger` enter the following command:

```sh
kubectl -n default apply -f - << END
apiVersion: eventing.knative.dev/v1alpha1
kind: Trigger
metadata:
  name: foo-display
spec:
  filter:
    sourceAndType:
      type: foo
  subscriber:
    ref:
     apiVersion: v1
     kind: Service
     name: foo-display
END
```

> Take notice of the attributes of the `Filter`. Every valid `CloudEvent` has attributes named `type` and `Source`. Triggers allow you to specify interest in specific `CloudEvents` by matching its `type` and `source`. 

Reviewer 3: What's a 'CloudEvent' vs a Knative event? Since we start using CloudEvents throughout the rest of the guide, we should take time to explain what that is to the reader _or_ avoid that topic and just talk about knative events?

Reviewer 2: Good point! Let's add in a link to https://github.com/cloudevents/spec and explain that CloudEvents are a particular spec for events on first use.

> Take notice of the attributes of the `Filter`. Every valid `CloudEvent` has attributes named `Type` and `Source`. Triggers allow you to specify interest in specific `CloudEvents` by matching its `Type` and `Source`. 

Reviewer 2: We should not capitalize things that aren't capitalized in the .yaml, like type, source, and filter.

> In this case, the `Trigger` you created searches for all `CloudEvents` of type `foo` and sends them to the `event consumer` `foo-display`.

Reviewer 2: I think you could also call foo-display a "subscriber." We could ask Akash when we should say "event consumer" and when we should say "subscriber". I do like that subscriber is a term that shows up in the .yaml, so I feel like that could be the clearest term for users, but I don't know for sure.

Author: @akashrv and I talked about this. Because of concerns with users mistaking "subscriber" in this context and other types of subscribers, it's best to keep it as event consumer globally.

```sh
kubectl -n default logs -l app=bar-display
```

The command shows the `Attributes` and `Data` of the events you sent to `bar-display`:

```sh
☁️  cloudevents.Event
 Validation: valid
 Context Attributes,
   specversion: 0.2
   type: anything-but-foo
   source: bar
   id: should-be-seen-by-bar
   time: 2019-05-20T17:59:49.044926148Z
   contenttype: application/json
 Extensions,
   knativehistory: default-broker-srk54-channel-24gls.default.svc.cluster.local
 Data,
   {
     "msg": "Hello World event from Knative - Bar"
   }
 ☁️  cloudevents.Event
 Validation: valid
 Context Attributes,
   specversion: 0.2
   type: foo
   source: bar
   id: should-be-seen-by-test
   time: 2019-05-20T17:59:54.211866425Z
   contenttype: application/json
 Extensions,
   knativehistory: default-broker-srk54-channel-24gls.default.svc.cluster.local
 Data,
   {
     "msg": "Hello World event from Knative - Both"
   } 
```

These results should match up with the CloudEvents you created in **Sending CloudEvents to the Broker**. If they do not, see the [Debugging Guide](./debugging/README.md) for more information.

## Cleaning up

Delete the namespace to conserve resources:

```sh

> kubectl delete namespace default

Reviewer 3: I don't think we should recommend deleting the default namespace. If we're going to use namespace deletion to simplify clean up, we should have the guide create a new namespace for the example.

> ## Cleaning up

Reviewer 2: Is this all we need to do to clean up? A lot of the install guides had pretty involved uninstall processes. We might ask Akash how to 100% uninstall eventing.

Reviewer 5: Generic uninstall eventing instructions should be a separate document so it can be linked to from every sample. It's more complicated than `kubectl delete -f eventing.yaml` because there are ordering requirements.

```

## What’s next 

You've learned the basics of the Knative Eventing workflow. Here are some additional resources to help you continue to build with the Knative Eventing component.

- [Using event importer to consume events](TODO)

> - [Configure `Broker` to Use a Different Channel](https://github.com/Harwayne/knative-eventing/blob/sample-3/configure-broker-to-use-a-different-channel.md)

Reviewer 2: For the What's Next section, we don't want to leave in any ToDos and we also don't want to link out to Adam's personal GitHub. We can link to existing content in our documentation or point to how-to guides for other components instead.

---
title: "Getting Started with Eventing"
#linkTitle: "OPTIONAL_ALTERNATE_NAV_TITLE"
weight: 9
type: "docs"
---

This getting started guide will walk you through all the steps involved in a typical Knative use case. You will learn how to create, send, and verify events in Knative.

## Before you begin

You need:

- A Kubernetes cluster with [Knative Eventing installed](./install/index.html). 

> ## Setting up Eventing subcomponents

Author: Would Trigger setup also go under the "Eventing Subcomponents" category, since (I think) it's an Eventing subcomponent? Currently, that section's placement is between event producers and event consumers.

Maybe an easier solution would be to have a _Setting Up_ heading with Setting Up Eventing Subcomponents, Setting Up Event Consumers, Setting Up Triggers, and Setting Up Event Producers as sub-headings, but that sounds a bit redundant.

What are your thoughts?

> ## Setting up Eventing subcomponents

Reviewer 5: Those 4 subheadings in that order seem good to me (and not redundant). Please note that as a major contributor my understanding of the concepts is not necessarily what the reader of this document will understand, so feel free to use your better judgment and ignore my suggestion. :smiley:

Before you start to send events, you need to create the components needed to transport the events. This guide walks you through creating each Eventing subcomponent.

> ### Configuring namespace for eventing

Reviewer: After performing the steps I think the title should change a bit to align with the actual task being performed.

```suggestion
### Creating and configuring an Eventing namespace
```

@grantr is it correct to define "an Eventing namespace" as a namespace that has the `knative-eventing-injection` label?

Reviewer 5: :+1: For this document "Eventing namespace" is a reasonable shorthand.

![technically-correct](https://user-images.githubusercontent.com/680/62226566-3a28b980-b36f-11e9-8cf9-36ffc472003f.png)

In most Knative tutorials, you will operate inside of a namespace. A namespace groups together subcomponents so that you can easily organize them.

1. Create the namespace. This command creates a namespace called **event-example**:

> ```sh

Reviewer: Each code block and any text must be nested within/under the list element with 4 spaces otherwise it renders poorly on the site (and breaks/resets numbers in the lists). Note: Our Hugo doc build only likes spaces (not tabs).
```suggestion
    ```sh
```

> kubectl create namespace event-example

Reviewer: ```suggestion
    kubectl create namespace event-example
```

> ```

Reviewer: ```suggestion
    ```
```

> 2. Set up the namespace for Knative Eventing. To set up the namespace, add a label to your namespace with this command:

Reviewer: Similar comment as above
```suggestion
2. Run the following command to add the 'knative-eventing-injection` label to your `event-example` namespace:
```

```sh
kubectl label namespace event-example knative-eventing-injection=enabled
```

> The `knative-eventing-injection` label triggers Knative to add Knative Eventing subcomponents such as the `Broker` in your namespace.

Reviewer: This is out of place and should be mentioned/explained before actually running the commands (provide the user the context for why to perform something, and then instruct them about how to do it)

> In the next section, you'll learn more about the `Broker`.

Reviewer 6: next section?

> In the next section, you'll learn more about the `Broker`.

Reviewer: This is true but I think whats more important is that you explain the relationship with what we have done, to whats next. For example, "Now that you have the `event-example` namespace, you can use it to create your Eventing subcomponents."

### Validating that the `Broker` is running

> The `Broker` ensures that every event sent by event producers is sent to the correct event consumers. While you created the `Broker` when you labeled your namespace as ready for eventing, it is important to verify that your `Broker` is working correctly. In this guide, you will use the default broker.

Reviewer: I dont think saying "you created the broker" is clear, since adding the label to a namespace "triggers the automatic creation of the Broker in that namespace".

Reviewer 2: Maybe "The Broker was created when you labeled the namespace as ready for eventing, but it's important to verify..."?

1. Use the following command to verify that the `Broker` is in a healthy state:

```sh
kubectl -n event-example get Broker default
```

This should return the following:

```sh
NAME      READY   REASON   HOSTNAME                                                           AGE
default   True             default-Broker.event-example.svc.cluster.local   1m
```

When the `Broker` is **READY**, it can begin to manage the events it receives.

2. If the **READY** column reads **False**, wait 2 minutes and repeat Step 1. If the **READY** column still reads **False**, see the [Debugging Guide](./debugging/README.md) to troubleshoot the issue.

Now your `Broker` is ready to manage your events.

### Creating event consumers

> These subcomponents receive the events sent by event producers (you'll create those a little later). You'll create two event consumers, `hello-display` and `goodbye-display`, so that you can see how to selectively send events to different consumers later.

Reviewer 5: Minor nit: I think I prefer the phrase "deliver events" over "send events". In my head, the producer sends events and the Broker delivers events. I could be the outlier here.

1. To create the `hello-display` subcomponent, enter the following command:

```sh
kubectl -n event-example apply -f - << END
apiVersion: apps/v1
kind: Deployment
metadata:
  name: hello-display
spec:
  replicas: 1
  selector:
    matchLabels: &labels
      app: hello-display
  template:
    metadata:
      labels: *labels
    spec:
      containers:
        - name: event-display

>           # Source code: https://github.com/knative/eventing-contrib/blob/release-0.6/cmd/event_display/main.go

Reviewer 7: have these images been updated for 0.7 release?

Author: When I spoke with @akashrv, he believed that was the current image. I'll go ahead and confirm if this is still the case.

Reviewer 6: It is an image of event-display container. It shouldn't be affected by Knative eventing 0.7 release and should work across releases.

---

# Service pointing at the previous Deployment. This will be the target for event
# consumption.
kind: Service
apiVersion: v1
metadata:
  name: goodbye-display
spec:
  selector:
    app: goodbye-display
  ports:
  - protocol: TCP
    port: 80
    targetPort: 8080
END
```

3. Just like the `Broker`, verify that the event consumers are working with the following command:

```sh
kubectl -n event-example get deployments hello-display goodbye-display
```

This commmand shows all of the deployments that have been created in `hello-display` and `goodbye-display`. You should have one replica in each event consumer:

```sh
NAME           DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
hello-display    1         1         1            1           26s
NAME           DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
goodbye-display    1         1         1            1           16s
```

The number of replicas in the **DESIRED** column should match the number of replicas in the **AVAILABLE** column. This may take a few minutes. If after two minutes the numbers still do not match the example, then see the [Debugging Guide](./debugging/README.md).

### Creating `Triggers`

> A `Trigger` expresses a workload's interest in events it wants to receive. A `Trigger` is split into two parts: the `Filter`, which tracks interested events, and the `Subscriber`, which determines where the event should be sent. Triggers can search for various types of events:

Reviewer 5: I would use "consumer" or "subscriber" instead of "workload" here, since you're using those words elsewhere (but maybe I missed a prior use of workload).

I'd also say the Filter "specifies interesting events". Filter and Subscriber probably should not be capitalized.

Use "match" instead of "search". Match is used elsewhere.

1. To create the first `Trigger` enter the following command:

```sh
kubectl -n event-example apply -f - << END
apiVersion: eventing.knative.dev/v1alpha1
kind: Trigger
metadata:
  name: hello-display
spec:
  filter:
    sourceAndType:
      type: greeting
  subscriber:
    ref:
     apiVersion: v1
     kind: Service
     name: hello-display
END
```

> Take notice of the attributes of the `Filter`.  All Knative events use a specification for describing event data called [`CloudEvent`](https://cloudevents.io/). Every valid `CloudEvent` has attributes named `type` and `source`. Triggers allow you to specify interest in specific `CloudEvents` by matching its `type` and `source`. 

Reviewer 7: in the first instance, I think it should read CloudEvents (with an s on the end).

In the 3rd sentence, 'Every valid 'CloudEvent' - singular is fine there.

> Take notice of the attributes of the `Filter`.  All Knative events use a specification for describing event data called [`CloudEvents`](https://cloudevents.io/). Every valid `CloudEvent` has attributes named `type` and `source`. Triggers allow you to specify interest in specific `CloudEvents` by matching its `type` and `source`. 

Reviewer 5: I think the only words needing to be code-blocked in this paragraph are `type` and `source`.

"matching their `type` and `source` attributes" instead of "matching its `type` and `source`".

> In this case, the `Trigger` you created searches for all `CloudEvents` of `type` `greeting` and sends them to the event consumer `hello-display`.

Reviewer 5: "matches" instead of "searches".

2. To add the second `Trigger`, enter the following command:

```sh
kubectl -n event-example apply -f - << END
apiVersion: eventing.knative.dev/v1alpha1
kind: Trigger
metadata:
  name: goodbye-display
spec:
  filter:
    sourceAndType:
      source: sendoff
  subscriber:
    ref:
     apiVersion: v1
     kind: Service
     name: goodbye-display
END
```

> Here, the command creates a `Trigger` that searches for all `CloudEvents` of `source` `sendoff` and sends them to the event consumer `goodbye-display`.

Reviewer 5: "matches" instead of "searches" here too.

3. Verify that the `Triggers` are running correctly with the following command:

```sh
kubectl -n event-example get triggers
```

> This command displays the **NAME**, **Broker**, **Subscriber_URI**, **AGE** and readiness of the `Triggers` in your namespace. You should see something like this:

Reviewer 5: This should probably be **SUBSCRIBER_URI**. Search and replace error?

```sh
NAME                 READY   REASON   BROKER    SUBSCRIBER_URI                                                                 AGE
goodbye-display          True             default   http://goodbye-display.event-example.svc.cluster.local/   9s
hello-display          True             default   http://hello-display.event-example.svc.cluster.local/    16s
```

> Both `Triggers` should be ready and pointing to the correct `Broker`  and `Subscriber_URI`. If this is not the case, see the [Debugging Guide](./debugging/README.md).

Reviewer 5: SUBSCRIBER_URI here too.

```sh
kubectl -n event-example apply -f - << END
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: curl
  name: curl
spec:
  containers:
    # This could be any image that we can SSH into and has curl.
  - image: radial/busyboxplus:curl
    imagePullPolicy: IfNotPresent
    name: curl
    resources: {}
    stdin: true
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    tty: true
END
```

You will use this `Pod` to send events in the next section.

## Sending CloudEvents to the `Broker`

Now that the `Pod` is created, you can create a `CloudEvent` by sending an HTTP request to the `Broker`. 

1. SSH into the `Pod` with the following command:

```sh
kubectl -n default attach curl -it
```

Now, you can make a HTTP request. To show the various types of events you can send, you will make three requests.

1. To make the first request, enter the following in the SSH terminal:

```sh
curl -v "default-broker.event-example.svc.cluster.local" \
  -X POST \
  -H "Ce-Id: say-hello" \
  -H "Ce-Specversion: 0.2" \
  -H "Ce-Type: greeting" \
  -H "Ce-Source: not-sendoff" \
  -H "Content-Type: application/json" \
  -d '{"msg":"Hello Knative!"}'
```

> This creates a `CloudEvent` called `say-hello` which has the `type` `greeting`. When an event is sent to the `Broker`, the `Trigger` `hello-display` will activate and send it to the event consumer of the same name.

Reviewer 5: I think I would not mention the id value here and just say, "This creates a CloudEvent that has the type `greeting`." (same with the next few examples). If you need to mention the id to reference it in later instructions, say, "This creates a CloudEvent that has the id `say-hello` and the type `greeting`."

 Context Attributes,
   specversion: 0.2
   type: not-greeting
   source: sendoff
   id: say-goodbye
   time: 2019-05-20T17:59:49.044926148Z
   contenttype: application/json
 Extensions,
   knativehistory: default-broker-srk54-channel-24gls.event-example.svc.cluster.local
 Data,
   {
     "msg": "Goodbye Knative!"
   }
 ☁️  cloudevents.Event
 Validation: valid
 Context Attributes,
   specversion: 0.2
   type: greeting
   source: sendoff
   id: say-hello-goodbye
   time: 2019-05-20T17:59:54.211866425Z
   contenttype: application/json
 Extensions,
   knativehistory: default-broker-srk54-channel-24gls.event-example.svc.cluster.local
 Data,
   {
     "msg": "Hello Knative! Goodbye Knative!"
   } 
```

These results should match up with the events you created in **Sending CloudEvents to the Broker**. If they do not, see the [Debugging Guide](./debugging/README.md) for more information.

## Cleaning up

Delete the namespace to conserve resources:

```sh
kubectl delete namespace event-example
```

## What’s next 

You've learned the basics of the Knative Eventing workflow. Here are some additional resources to help you continue to build with the Knative Eventing component.

> - [Knative Eventing Overview](./README.md)

Reviewer 6: As per the [draft documentation](https://docs.google.com/document/d/1nxNNzHoVoWulibSZBFGHo8EjVk47FK55NtfcBDto8w0/), the plan was to have a Getting started section with following topics:
1. Quick Start: Deploy "Hello World" eventing solution
2. Step-by-step: Deploy "Hello World" eventing solution --> This is the current PR
3. Using event importer to consume events
4. Configure Broker to use a different Channel

Question: Above 4 topics have a gradual flow (continuation) from #1 to #4. Is there a plan to maintain that flow and create more pages and have pre-req and next steps?

Reviewer 2: @akashrv  I'd think of this PR as 1 + 2 merged together. We are not currently planning to create the stripped-down, no explanation version of this guide that does all this work in one step. 

For #3, I am beginning to work on that. I've got a meeting with Ville tomorrow.

#4 is further out at this point; I don't have a timeline for when that will be covered.

When #3 exists, it will definitely be linked to from this guide as a good next step, and same for #4 further down the road.

---
title: "Getting Started with Eventing"
linkTitle: "Getting Started"
weight: 9
type: "docs"
---

Use this guide to learn how to learn how to create, send, and verify events in Knative. The steps in this guide demonstrate a basic developer flow for managing events in Knative, including:

- [Installing the Knative Eventing component](#installing-knative-eventing)
- [Creating and configuring the resources needed to manage events](#setting-up-knative-eventing-resources)
- [Creating and sending events with HTTP requests](#sending-cloudevents-to-the-broker)
- [Verifying events were sent correctly](#verifying-events-were-received)

> 

Author: Per our discussion yesterday, I am removing references to the Eventing over until some issues in that document are fixed.

### Creating `Triggers`

A `Trigger` expresses an event consumer's interest in events it wants to receive. A `Trigger` is split into two parts: the `filter`, which specifies interesting events, and the `subscriber`, which determines where the event should be sent. `Triggers` can match various types of events:

1. To create the first `Trigger` run the following command:

    ```sh
    kubectl -n event-example apply -f - << END
    apiVersion: eventing.knative.dev/v1alpha1
    kind: Trigger
    metadata:
      name: hello-display
    spec:
      filter:
        sourceAndType:
          type: greeting
      subscriber:
        ref:
         apiVersion: v1
         kind: Service
         name: hello-display
    END
    ```

    Take notice of the attributes of the Filter.  All Knative events use a specification for describing event data called [CloudEvents](https://cloudevents.io/). Every valid CloudEvent has attributes named `type` and `source`. Triggers allow you to specify interest in specific CloudEvents by matching its `type` and `source`. 

>     In this case, the `Trigger` you created matches all `CloudEvents` of `type` `greeting` and sends them to the event consumer `hello-display`.

Reviewer: ```suggestion
    Result: You created a `Trigger` that sends all events of type `greeting` to your event consumer named `hello-display`.
```

3. Verify that the `Triggers` are working correctly by running the following command:

    ```sh
    kubectl -n event-example get triggers
    ```

    This command displays the **NAME**, **Broker**, **SUBSCRIBER_URI**, **AGE** and readiness of the `Triggers` in your namespace. You should see something like this:

    ```sh
    NAME                 READY   REASON   BROKER    SUBSCRIBER_URI                                                                 AGE
    goodbye-display          True             default   http://goodbye-display.event-example.svc.cluster.local/   9s
    hello-display          True             default   http://hello-display.event-example.svc.cluster.local/    16s
    ```

    Both `Triggers` should be ready and pointing to the correct **Broker**  and **SUBSCRIBER_URI**. If this is not the case, see the [Debugging Guide](./debugging/README.md).

You have now created all of the subcomponents needed to recieve and manage events. In the next section, you will make the subcomponent that will be used to create your events.

### Creating event producers

> Since this guide uses manual curl requests to send events, you will need to make an event producer called a `Pod`. The `Broker` is only exposed from within the Kubernetes cluster, so you will run the curl request from the `Pod`.

Reviewer 5: > a `Pod` that forwards your `curl` commands as events to the `Broker`

The word forward here is confusing. It makes more sense to me to use the word execute. For example:

> In the following step, the event producer that you create is a `Pod` that executes your curl commands which send events to the `Broker` in your Eventing cluster.

Could also replace "send" with submit".

> Since this guide uses manual curl requests to send events, you will need to make an event producer called a `Pod`. The `Broker` is only exposed from within the Kubernetes cluster, so you will run the curl request from the `Pod`.

Reviewer 5: Otherwise accurate :+1:

      -H "Ce-Specversion: 0.2" \
      -H "Ce-Type: greeting" \
      -H "Ce-Source: sendoff" \
      -H "Content-Type: application/json" \
      -d '{"msg":"Hello Knative! Goodbye Knative!"}'
   ```

  This creates a `CloudEvent` that has the `type` `greeting` and the`source` `sendoff`. When the event is sent to the `Broker`, the `Triggers` `hello-display` and `goodbye-display` will activate and send it to the event consumer of the same name.

  If the event has been received, you should receive a `202 Accepted` response.

4. Exit SSH.

If everything has been done correctly, you should have sent 2 `CloudEvents` to the `hello-display` event consumer and 2 `CloudEvents` to the `goodbye-display` event consumer (note that `say-hello-goodbye` is sent to *both* `hello-display` and `goodbye-display`). You will verify that these events were received correctly in the next section.

## Verifying events were received {:#verify}

After sending events, verify that the events were received by the appropriate `Subscribers`.

1. Look at the logs for the `hello-display` event consumer by running the following command:

    ```sh
    kubectl -n event-example logs -l app=hello-display
    ```

>     The command shows the `Attributes` and `Data` of the events you sent to `hello-display`:

Reviewer: @grantr I don't get the expected results that we documented here.

I only ever get the last event with far less detail:

```
[rescarez] ~/knative/docs (master)
$ kubectl -n event-example logs -l app=hello-display
  source: sendoff
  id: say-hello-goodbye
  time: 2019-08-12T19:49:09.7953344Z
  contenttype: application/json
Extensions,
  knativehistory: default-kn-trigger-channel-cqjls.event-example.svc.cluster.local
Data,
  {
    "msg": "Hello Knative! Goodbye Knative!"
  }

[rescarez] ~/knative/docs (master)
$ kubectl -n event-example logs -l app=goodbye-display
  source: sendoff
  id: say-hello-goodbye
  time: 2019-08-12T19:49:09.7953344Z
  contenttype: application/json
Extensions,
  knativehistory: default-kn-trigger-channel-cqjls.event-example.svc.cluster.local
Data,
  {
    "msg": "Hello Knative! Goodbye Knative!"
  }

```

So if I rerun the first two curl commands, i see:

```
[rescarez] ~/knative/docs (master)
$ kubectl -n event-example logs -l app=hello-display
  source: not-sendoff
  id: say-hello
  time: 2019-08-12T20:25:46.104264737Z
  contenttype: application/json
Extensions,
  knativehistory: default-kn-trigger-channel-cqjls.event-example.svc.cluster.local
Data,
  {
    "msg": "Hello Knative!"
  }

[rescarez] ~/knative/docs (master)
$ kubectl -n event-example logs -l app=goodbye-display
  source: sendoff
  id: say-goodbye
  time: 2019-08-12T20:25:12.895347894Z
  contenttype: application/json
Extensions,
  knativehistory: default-kn-trigger-channel-cqjls.event-example.svc.cluster.local
Data,
  {
    "msg": "Goodbye Knative!"
  }

```

Reviewer 5: @RichieEscarez `kubectl logs` when used with a selector (`-l app=goodbye-display`) only shows the 10 most recent log lines. You can change this with the `--tail=N` option which specifies the number of lines to display. Alternately `--tail=-1` displays all lines.

```
kubectl logs -l app=goodbye-display --tail=100
```

Author: Thanks Grant! I'll use `--tail=-1` just to cover everything.

---
title: "Getting Started with Eventing"
linkTitle: "Getting Started"
weight: 9
type: "docs"
---

> Use this guide to learn how to learn how to create, send, and verify events in Knative. The steps in this guide demonstrate a basic developer flow for managing events in Knative, including:

Reviewer 2: "to learn how to learn how to" --> "to learn how to"

- [Installing the Knative Eventing component](#installing-knative-eventing)
- [Creating and configuring Knative Eventing Resources](#setting-up-knative-eventing-resources)
- [Sending events with HTTP requests](#sending-events-to-the-broker)
- [Verifying events were sent correctly](#verifying-events-were-received)

## Before you begin

To complete this guide, you will need the following installed and running:

- A [Kubernetes cluster](https://kubernetes.io/docs/concepts/cluster-administration/cluster-administration-overview/) running v1.11 or higher
- [Kubectl CLI tool](https://kubernetes.io/docs/reference/kubectl/overview/) v1.10 or higher
- [curl v7.65 or higher](https://curl.haxx.se/download.html)
- Knative Eventing Component

### Installing Knative Eventing 

If you previously [created a Knative cluster](../install/_index.md), you might already have Knative Eventing installed and running. You can check to see if the Eventing component exists on your cluster by running:

```sh
kubectl get pods --namespace knative-eventing
```

> 

Reviewer: Per the error i mention in another comment, we need to clarify that the "in-memory channel provisioner" is required and must be listed.

> If the `knative-eventing` namespace does not exist, use the follow steps to add the Knative Eventing component to your cluster:

Reviewer: ```suggestion
Result:

<code>shell
NAME                                            READY   STATUS    RESTARTS   AGE
eventing-controller-6cdf6ff785-44k98            1/1     Running   0          29m
eventing-webhook-6c4f6699d8-qclbx               1/1     Running   0          29m
imc-controller-85cdb4946b-h2msj                 1/1     Running   0          13m
imc-dispatcher-5f4689d868-fspt6                 1/1     Running   0          13m
in-memory-channel-controller-d9db9d879-f2jw6    1/1     Running   0          13m
in-memory-channel-dispatcher-79bc7f46cd-4mvj6   1/1     Running   0          13m
sources-controller-5847564f4f-z59xc             1/1     Running   0          29m
</code>

If the `knative-eventing` namespace does not exist or the `in-memory-channel-controller-*` is not listed, use the following steps to install Knative Eventing with the in-memory channel provisioner:
```

NOTE: replace the `<code>` tags with markdown code blocks (the inline editor does not like the nested tick marks)

         kind: Service
         name: hello-display
    END
    ```

   Result: You created a `Trigger` that sends all events of type `greeting` to your event consumer named `hello-display`.

2. To add the second `Trigger`, run the following command:

    ```sh
    kubectl --namespace event-example apply --filename - << END
    apiVersion: eventing.knative.dev/v1alpha1
    kind: Trigger
    metadata:
      name: goodbye-display
    spec:
      filter:
        sourceAndType:
          source: sendoff
      subscriber:
        ref:
         apiVersion: v1
         kind: Service
         name: goodbye-display
    END
    ```

   Result: You created a `Trigger` that sends all events of source `sendoff` to your event consumer named `goodbye-display`.

3. Verify that the `Triggers` are working correctly by running the following command:

    ```sh
    kubectl --namespace event-example get triggers
    ```

    Result: Both the `hello-display` and `goodbye-display` `Triggers` that you created should be listed, for example:

    ```sh
    NAME                 READY   REASON   BROKER    SUBSCRIBER_URI                                                                 AGE

>     goodbye-display          True             default   http://goodbye-display.event-example.svc.cluster.local/   9s

Reviewer: fix column alignment

```suggestion
    goodbye-display        True             default   http://goodbye-display.event-example.svc.cluster.local/   9s
```

    ```sh
    kubectl --namespace event-example logs -l app=hello-display --tail=-1
    ```

    Result: The command shows the `Attributes` and `Data` of the events you sent to `hello-display`:

    ```sh
    ☁️  cloudevents.Event
    Validation: valid
    Context Attributes,
      specversion: 0.2
      type: greeting
      source: not-sendoff
      id: say-hello
      time: 2019-05-20T17:59:43.81718488Z
      contenttype: application/json
    Extensions,
      knativehistory: default-broker-srk54-channel-24gls.event-example.svc.cluster.local
    Data,
      {
        "msg": "Hello Knative!"
      }
    ☁️  cloudevents.Event
    Validation: valid
    Context Attributes,
      specversion: 0.2
      type: greeting
      source: sendoff
      id: say-hello-goodbye
      time: 2019-05-20T17:59:54.211866425Z
      contenttype: application/json
    Extensions,
      knativehistory: default-broker-srk54-channel-24gls.event-example.svc.cluster.local
    Data,
     {
       "msg": "Hello Knative! Goodbye Knative!"
     }
    ```

2. Look at the logs for the `goodbye-display` event consumer by running the following command:

    ```sh

>     kubectl --namespace event-example logs -l app=goodbye-display --tail=-1

Reviewer: These commands still only show me that "10 line" shortened response. Whats missing so that I see the same verbose examples that are shown here?

Author: I changed the --tail flag to show 100 lines, and it was more verbose. According to kubectl documentation, tail=-1 is supposed to show all, and it's not working for me again. But tail=100 works, so I'll keep that for now.

   ```

## Setting up Knative Eventing Resources 

Before you start to manage events, you need to create the objects needed to transport the events. 

### Creating and configuring an Eventing namespace

In this section you create the `event-example` namespace and then add the `knative-eventing-injection` label to that namespace. You use namespaces to group together and organize your Knative resources, including the Eventing subcomponents.

1. Run the following command to create a namespace called `event-example`:

    ```sh
    kubectl create namespace event-example
    ```

    This creates an empty namespace called `event-example`.

2. Add a label to your namespace with the following command:

    ```sh
    kubectl label namespace event-example knative-eventing-injection=enabled
    ```

   This gives the `event-example`  namespace the `knative-eventing-injection` label, which adds resources that will allow you to manage your events.

In the next section, you will need to verify that the resources you added in this section are running correctly. Then, you can create the rest of the eventing resources you need to manage events.

### Validating that the `Broker` is running

> The `Broker` ensures that every event sent by event producers arrives at the correct event consumers. The `Broker` was created when you labeled your namespace as ready for eventing, but it is important to verify that your `Broker` is working correctly. In this guide, you will use the default broker.

Reviewer 6: link to https://github.com/knative/docs/blob/master/docs/eventing/broker-trigger.md#Broker

    ```sh
    kubectl --namespace event-example get deployments hello-display goodbye-display
    ```

    This lists the `hello-display` and `goodbye-display` consumers that you deployed:

    ```sh
    NAME           DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
    hello-display    1         1         1            1           26s
    NAME           DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
    goodbye-display    1         1         1            1           16s
   ```

    The number of replicas in your **DESIRED** column should match the number of replicas in your **AVAILABLE** column, which might take a few minutes. If after two minutes the numbers do not match, then see the [Debugging Guide](./debugging/README.md) to help troubleshoot the issue.

### Creating `Triggers`

A [Trigger](./broker-trigger.md#trigger) defines the events that you want each of your event consumers 
to receive. Your `Broker` uses triggers to forward events to the right consumers. Because all events 
in Knative align with the [CloudEvents](https://cloudevents.io/) specification, each event in Knative must include 

> the `type` and `source` attributes. Each trigger uses the `type` and `source` attributes to match events 

Reviewer 6: This is no longer true (even though it was when you started to write this document :))
Change the text to:
"All events in Knative align with the [CloudEvents](https://cloudevents.io/) specification. Each Trigger can specify a filter to select relevant events based on the [Cloud Event context attributes](https://github.com/cloudevents/spec/blob/master/spec.md#context-attributes).

to consumers, where a trigger's `filter` specifies the event `type` attribute, and a trigger's `subscriber` defines the event `source` attribute.

1. To create the first `Trigger`, run the following command:

    ```sh
    kubectl --namespace event-example apply --filename - << END
    apiVersion: eventing.knative.dev/v1alpha1
    kind: Trigger
    metadata:
      name: hello-display
    spec:
      filter:

>         sourceAndType:

Reviewer 6: replace 'sourceAndType' with 'attributes'

          type: greeting
      subscriber:
        ref:
         apiVersion: v1
         kind: Service
         name: hello-display
    END
    ```

   The command creates a trigger that sends all events of type `greeting` to your event consumer named `hello-display`.

2. To add the second `Trigger`, run the following command:

    ```sh
    kubectl --namespace event-example apply --filename - << END
    apiVersion: eventing.knative.dev/v1alpha1
    kind: Trigger
    metadata:
      name: goodbye-display
    spec:
      filter:

>         sourceAndType:

Reviewer 6: replace 'sourceAndType' with 'attributes'

    ```sh
    kubectl create namespace event-example
    ```
    Result: You have created an empty namespace called `event-example`.

2. Add a label to your namespace with the following command:

    ```sh
    kubectl label namespace event-example knative-eventing-injection=enabled
    ```

   Result: The `event-example`  namespace now has the `knative-eventing-injection` label. This label adds resources that will allow you to manage your events.

In the next section, you will need to verify that the resources you added in this section are running correctly. Then, you can create the rest of the eventing resources you need to manage events.

### Validating that the `Broker` is running

The `Broker` ensures that every event sent by event producers is sent to the correct event consumers. The `Broker` was created when you labeled your namespace as ready for eventing, but it is important to verify that your `Broker` is working correctly. In this guide, you will use the default broker.

1. Run the following command to verify that the `Broker` is in a healthy state:

    ```sh
    kubectl --namespace event-example get Broker default
    ```

>     Result: The `Broker` that you created should be listed, for example:

Reviewer 2: Try to avoid using "should." See [the entry for "should"](https://developers.google.com/style/word-list#letter-S) in the Google Developer world list.

To also remove "Result:", I'd suggest something like:

> This returns information for the `Broker` you just created:

    ```sh
    NAME                   READY   REASON   BROKER    SUBSCRIBER_URI                                                                 AGE
    goodbye-display        True             default   http://goodbye-display.event-example.svc.cluster.local/                        9s
    hello-display          True             default   http://hello-display.event-example.svc.cluster.local/                          16s
    ```

    Both triggers should be ready and pointing to the correct **Broker** (your triggers should be pointing at the default broker)  and **SUBSCRIBER_URI** (your triggers should be pointing to triggerName.namespaceName.svc.cluster.local). If this is not the case, see the [Debugging Guide](./debugging/README.md) to help troubleshoot the issue.

You have now created all of the resources needed to receive and manage events. You created the `Broker`, which manages the events sent to event consumers with the help of triggers.  In the next section, you will make the event producer that will be used to create your events.

### Creating event producers

> In this section you will create an event producer that you can use to interact with the Knative Eventing subcomponents you created earlier.

Reviewer 2: It looks like you have some paragraphs here where you've started a new line at the end, but you need one empty line between paragraphs for proper markdown style.

> Typically, events are created systematically but in this guide `curl` is used to allow you to manually send individual events and demonstrate how each of those events are received by the correct event consumer within your Knative Eventing cluster.

Reviewer 2: There's some passive voice throughout this doc. A good way to check for it is to do a search for " is " and " are ". I'll try to flag it where I find it and offer suggestions.

For example, you might rewrite line 298 this way to avoid the passive voice of "is used" and to remove extra words:

> Most events are created systematically, but this guide uses `curl` to manually send individual events and demonstrate how these events are received by the correct event consumer.

> Since the `Broker` is accessible only from within your Eventing cluster, you must create a `Pod` within that cluster to act as your event producer.  

Reviewer 2: To avoid the passive voice, line 299 could be rewritten as:

> Because you can only access the `Broker` from within your Eventing cluster, you must create a `Pod` within that cluster to act as your event producer.

In the following step, you will create a `Pod` that executes your `curl` commands to send events to the `Broker` in your Eventing cluster.

 To create the `Pod`, run the following command:

```sh
kubectl --namespace event-example apply --filename - << END
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: curl
  name: curl
spec:
  containers:
    # This could be any image that we can SSH into and has curl.
  - image: radial/busyboxplus:curl
    imagePullPolicy: IfNotPresent
    name: curl
    resources: {}
    stdin: true
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    tty: true
END
```

> Now that your Eventing cluster is set up to send and consume events, you will use HTTP requests to manually send separate events and demonstrate how each of those events can target your individual event consumers in the next section.

Reviewer 2: To avoid the passive voice, line 327 could be rewritten as:

> Now that you've set up your Eventing cluster to send and consume events, ...

## Sending Events to the `Broker` 

> Now that the `Pod` is created, you can create an event by sending an HTTP request to the `Broker`. SSH into the `Pod` by running the following command:

Reviewer 2: To avoid the passive voice, line 331 could be rewritten as:

> Now that you've created the `Pod`, ...

```sh
  kubectl --namespace event-example attach curl -it
```

Result: You should be inside of the Pod, and you can now make a HTTP request. You should see this prompt:
    
```sh
    Defaulting container name to curl.
    Use 'kubectl describe pod/ -n event-example' to see all of the containers in this pod.
    If you don't see a command prompt, try pressing enter.
    [ root@curl:/ ]$
```

To show the various types of events you can send, you will make three requests.

1. To make the first request, which creates an event that has the `type` `greeting`, run the following in the SSH terminal:

    ```sh
    curl -v "default-broker.event-example.svc.cluster.local" \
      -X POST \
      -H "Ce-Id: say-hello" \
      -H "Ce-Specversion: 0.2" \
      -H "Ce-Type: greeting" \
      -H "Ce-Source: not-sendoff" \
      -H "Content-Type: application/json" \
      -d '{"msg":"Hello Knative!"}'
    ```

>     Result: When your event is sent to the `Broker`, `hello-display` will activate and send it to the event consumer of the same name.

Reviewer 2: To avoid the passive voice, line 361 could be rewritten as:

> When the `Broker` receives your event ...

    If the event has been received, you will receive a `202 Accepted` response similar to the one below:
    ```sh
    < HTTP/1.1 202 Accepted
    < Content-Length: 0
    < Date: Mon, 12 Aug 2019 19:48:18 GMT
    ```

2. To make the second request, which creates an event that has the `source` `sendoff`, run the following in the SSH terminal:

    ```sh
    curl -v "default-broker.event-example.svc.cluster.local" \
      -X POST \
      -H "Ce-Id: say-goodbye" \
      -H "Ce-Specversion: 0.2" \
      -H "Ce-Type: not-greeting" \
      -H "Ce-Source: sendoff" \
      -H "Content-Type: application/json" \
      -d '{"msg":"Goodbye Knative!"}'
    ```

>    Result: When your event is sent to the `Broker`, `goodbye-display` will activate and send the event to the event consumer of the same name.

Reviewer 2: Same recommendation as line 361.

>     If the event has been received, you will receive a `202 Accepted` response similar to the one below:

Reviewer 2: There should be a new line between 386 and 387.

    ```sh
    < HTTP/1.1 202 Accepted
    < Content-Length: 0
    < Date: Mon, 12 Aug 2019 19:48:18 GMT
    ```

> 

Reviewer 2: There's an extra new line here; line 392 can be deleted.

3. To make the third request, which creates an event that has the `type` `greeting` and the`source` `sendoff`, run the following in the SSH terminal:

    ```sh
    curl -v "default-broker.event-example.svc.cluster.local" \
      -X POST \
      -H "Ce-Id: say-hello-goodbye" \
      -H "Ce-Specversion: 0.2" \
      -H "Ce-Type: greeting" \
      -H "Ce-Source: sendoff" \
      -H "Content-Type: application/json" \
      -d '{"msg":"Hello Knative! Goodbye Knative!"}'
   ```

   Result: When your event is sent to the `Broker`,  `hello-display` and `goodbye-display` will activate and send the event to the event consumer of the same name.

   If the event has been received, you will receive a `202 Accepted` response similar to the one below:
    ```sh
    < HTTP/1.1 202 Accepted
    < Content-Length: 0
    < Date: Mon, 12 Aug 2019 19:48:18 GMT
    ```

4. Exit SSH by typing `exit` into the command prompt.
    

> If everything has been done correctly, you should have sent 2 events to the `hello-display` event consumer and 2 events to the `goodbye-display` event consumer (note that `say-hello-goodbye` is sent to *both* `hello-display` and `goodbye-display`). You will verify that these events were received correctly in the next section.

Reviewer 2: Spell out numbers smaller than ten. "2" -> "two". [Style guide reference](https://developers.google.com/style/numbers#numbers-as-words).

> If everything has been done correctly, you should have sent 2 events to the `hello-display` event consumer and 2 events to the `goodbye-display` event consumer (note that `say-hello-goodbye` is sent to *both* `hello-display` and `goodbye-display`). You will verify that these events were received correctly in the next section.

Reviewer 2: See above comment about "should."

I'd also cut out "If everything has been done correctly".

## Verifying events were received 

After sending events, verify that your events were received by the appropriate `Subscribers`.

1. Look at the logs for the `hello-display` event consumer by running the following command:

    ```sh
    kubectl --namespace event-example logs -l app=hello-display --tail=100
    ```

>     Result: The command shows the `Attributes` and `Data` of the events you sent to `hello-display`:

Reviewer 2: I'd suggest "This returns" instead of "The command shows"

    ```sh
    ☁️  cloudevents.Event
    Validation: valid
    Context Attributes,
      specversion: 0.2
      type: greeting
      source: not-sendoff
      id: say-hello
      time: 2019-05-20T17:59:43.81718488Z
      contenttype: application/json
    Extensions,
      knativehistory: default-broker-srk54-channel-24gls.event-example.svc.cluster.local
    Data,
      {
        "msg": "Hello Knative!"
      }
    ☁️  cloudevents.Event
    Validation: valid
    Context Attributes,
      specversion: 0.2
      type: greeting
      source: sendoff
      id: say-hello-goodbye
      time: 2019-05-20T17:59:54.211866425Z
      contenttype: application/json
    Extensions,
      knativehistory: default-broker-srk54-channel-24gls.event-example.svc.cluster.local
    Data,
     {
       "msg": "Hello Knative! Goodbye Knative!"
     }
    ```

2. Look at the logs for the `goodbye-display` event consumer by running the following command:

    ```sh
    kubectl --namespace event-example logs -l app=goodbye-display --tail=100
    ```

>     Result: The command shows the `Attributes` and `Data` of the events you sent to `goodbye-display`:

Reviewer 2: Same comment as line 429.

       time: 2019-05-20T17:59:49.044926148Z
       contenttype: application/json
     Extensions,
       knativehistory: default-broker-srk54-channel-24gls.event-example.svc.cluster.local
    Data,
       {
         "msg": "Goodbye Knative!"
       }
     ☁️  cloudevents.Event
     Validation: valid
     Context Attributes,
       specversion: 0.2
       type: greeting
       source: sendoff
       id: say-hello-goodbye
       time: 2019-05-20T17:59:54.211866425Z
       contenttype: application/json
     Extensions,
       knativehistory: default-broker-srk54-channel-24gls.event-example.svc.cluster.local
     Data,
      {
        "msg": "Hello Knative! Goodbye Knative!"
      } 
   ```

## Cleaning up

After you finish this guide, delete your namespace to conserve resources if you do not plan to use them.

Note: If you plan to continue learning about Knative Eventing with one of our [code samples](./samples/_index.md), check the requirements of the sample and make sure you do not need a namespace before you delete `event-example`. You can always reuse your namespaces.

Run the following command to delete `event-example`:

```sh
kubectl delete namespace event-example
```

> Result: The namespace is removed from your cluster, along with the resources inside the namespace.

Reviewer 2: To remove the passive voice, I'd suggested changing line 518 to:

> This removes the namespace and all of its resources from your cluster.
